### ISSUE TITLE
DSA signature not working on master

### ISSUE THREAD

**Issue by PipoloyJo** [sparkle-project/Sparkle#562]
Hi,

I'm currently integrating Sparkle in my project and I found an issue with DSA signing on master.

It works fine on the last release (1.10) but seems to be broken on HEAD. I got this message:
____
Sparkle: Public DSA keys differ or are absent, and both apps are not code signed. The update will be rejected.
Sparkle: Error: An error occurred while extracting the archive. Please try again later. The update is improperly signed. (URL (null))
____
From what I can see from git log, it seems to be related to Zorg branch, starting with this commit 28ea47f796b2f433720a40b4b4903f844c4b6967

I don't have much time to investigate this issue, but if I can help let me know. 

Really appreciate your work, thanks guys!

Pipo

**Comment by pornel** [comment]
Thanks for the report. 

Do you use DSA keys? Do you use Apple Code Signing?

**Comment by PipoloyJo** [comment]
I'm using DSA keys.

**Comment by zorgiepoo** [comment]
The relevant code is here: https://github.com/sparkle-project/Sparkle/blob/master/Sparkle/SUBasicUpdateDriver.m#L277

Which is to say, the code doesn't think that the DSA public keys in your old app and in your new app match, or the public DSA key is absent in one of these apps.

Can you verify that your DSA public key is properly specified in both the old and new app, and that they are the *same*?

**Comment by PipoloyJo** [comment]
As I said earlier, DSA code signing worked with 1.10 release. Steps I followed:

- I generated two packages of my app with incremental version number, both packaged with the same public DSA key
- I installed the first one on my machine, and placed the other one on the server, with the DSA key and the appcast properly updated (and the private DSA signature)
- From the installed one I ran the "Check for Update..." check which properly found that a newer version of my app was available
- After that Sparkle was able to shutdown my app using the Autoupdate.app launch the new one and place it back in Applications folder. Smooth and easy.

With master I got stuck after downloading the update. Same code, same app, same key. I can try to rebuild the package with more debugging log, and see what changed. I've already started to look in the SUBasicUpdateDriver code.

**Comment by zorgiepoo** [comment]
You will want to look why dsaKeysMatch BOOL is NO. (Is publicDSAKey nil? is newPublicDSAKey nil? Does [publicDSAKey isEqualToString:newPublicDSAKey] evaluate to NO?)

**Comment by zorgiepoo** [comment]
Also, for what it's worth, I've been able to create a test application and have it properly validated using only DSA keys. Having an update work in 1.10 is not sufficient evidence for something being wrong with master, since 1.10 also tolerated insecure updates more.

**Comment by PipoloyJo** [comment]
Ok, that was my next guess: some security upgrade.

I'll recompile the framework and check this out. Thanks for your quick answers, i'll report progress tomorrow.

**Comment by PipoloyJo** [comment]
Hi again, I've found out my issue, the remote package was not containing the DSA key anymore. The update package has been updated by our nightly build system and was compiled without shipping Sparkle. 

So the host app had the key but the update did not. I've replace the nightly package, tested again and it went well. 

It was a security leak in 1.10 since I was able to update a signed app with an unsecure app not containing a DSA key.

Thanks for your support!
